File: python/handler.py
import json
import os

import boto3
from boto3.dynamodb.conditions import Key


def hello(event, context):
    dynamodb = boto3.resource("dynamodb", region_name=os.environ["AWS_REGION"])
    table = dynamodb.Table(os.environ["DYNAMODB_TABLE"])

    # Fetch all items from the database by index.
    items = table.query(
        IndexName=os.environ["DYNAMODB_GSI_USERID_NOTEID"],
        ExpressionAttributeNames={
            "#note_text": "text"
        },
        ProjectionExpression="userId, noteId, notebook, #note_text",
        KeyConditionExpression=Key("userId").eq("arbiter"),
        ScanIndexForward=True
    )

    results = items["Items"] if "Items" in items else {}

    return {
        "statusCode": 200,
        "body": json.dumps({"results": results})
    }

File: typescript/handler.ts
'use strict'

import { APIGatewayProxyHandler } from 'aws-lambda';
import { DynamoDB, AWSError } from 'aws-sdk';
import 'source-map-support/register';
import { PromiseResult } from 'aws-sdk/lib/request';

export const hello: APIGatewayProxyHandler = async (_event, _context) => {

  let ddb = new DynamoDB({apiVersion: '2012-08-10', region: 'us-west-2'});
  const params = {
    TableName: process.env.DYNAMODB_TABLE,
    IndexName: process.env.DYNAMODB_GSI_USERID_NOTEID,
    KeyConditionExpression: 'userId = :userId',
    ProjectionExpression: 'userId, noteId, notebook, #note_text',
    ExpressionAttributeNames: {
      '#note_text': 'text'
    },
    ExpressionAttributeValues: {
      ':userId': { 'S': 'arbiter' }
    },
    ScanIndexForward: true
  };

  let response: PromiseResult<DynamoDB.QueryOutput, AWSError> =
    await ddb.query(params, function(err: AWSError, data: DynamoDB.QueryOutput) {
    if (err) {
      return err;
    } else {
      return data.Items;
    }
  }).promise();

  let unmarshalled = response.Items.map(element => {
    return DynamoDB.Converter.unmarshall(element);
  });

  return {
    statusCode: 200,
    body: JSON.stringify({
      results: unmarshalled
    }, null, 2),
  };
}
